### PR TITLE
PHOENIX-6271: Effective DDL generated by SchemaExtractionTool should …

### PR DESCRIPTION
**Pull request by swaroopak**
…maintain the order of PK and other columns

### CODE REVIEW

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 55s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.x Compile Tests _ |
| +0 :ok: |  mvndep  |   5m 28s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  11m 11s |  4.x passed  |
| +1 :green_heart: |  compile  |   1m 48s |  4.x passed  |
| +1 :green_heart: |  checkstyle  |   0m 58s |  4.x passed  |
| +1 :green_heart: |  javadoc  |   1m 12s |  4.x passed  |
| +0 :ok: |  spotbugs  |   3m 57s |  phoenix-core in 4.x has 946 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   0m 50s |  phoenix-tools in 4.x has 3 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 14s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   7m 16s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 54s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 54s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 45s |  phoenix-core: The patch generated 2 new + 719 unchanged - 3 fixed = 721 total (was 722)  |
| -1 :x: |  checkstyle  |   0m 16s |  phoenix-tools: The patch generated 11 new + 215 unchanged - 8 fixed = 226 total (was 223)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   1m  7s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   5m 31s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 241m 55s |  phoenix-core in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 10s |  phoenix-tools in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 20s |  The patch does not generate ASF License warnings.  |
|  |   | 292m 15s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/1/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1212 |
| JIRA Issue | PHOENIX-6271 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux c6050475172b 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.x / c3f166e |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/1/artifact/yetus-general-check/output/diff-checkstyle-phoenix-tools.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/1/testReport/ |
| Max. process+thread count | 4725 (vs. ulimit of 30000) |
| modules | C: phoenix-core phoenix-tools U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/1/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java [hunk 1]
```diff
@@ -1275,15 +1275,21 @@ public static String getPTableFullNameWithQuotes(String pSchemaName, String pTab
             pTableName = "\""+pTableName+"\"";
         }
         if(tableNameNeedsQuotes || schemaNameNeedsQuotes) {
-            pTableFullName = pSchemaName + "." + pTableName;
+            if (pSchemaName != null && !pSchemaName.isEmpty()) {
```
Comment (gokceni) [inline phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1278]: nit: Strings.isNullOrEmpty

Comment (gokceni) [reply phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1278]: Also do you need to think about namespaceenabled? I am guessing no

File: phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java [hunk 2]
```diff
@@ -1275,15 +1275,21 @@ public static String getPTableFullNameWithQuotes(String pSchemaName, String pTab
...
+            }
         }
-
         return pTableFullName;
     }
 
     private static boolean isQuotesNeeded(String name) {
         // first char numeric or non-underscore
-        if(!Character.isAlphabetic(name.charAt(0)) && name.charAt(0)!='_') {
+        if (name == null || name.isEmpty()) {
```
Comment (gokceni) [inline phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:1289]: nit: Strings.isNullOrEmpty

**Review by gokceni — COMMENTED** [review]
some nits

**Review by gjacoby126 — CHANGES_REQUESTED** [review]
Maybe I'm misunderstanding what you're trying to do in this patch, but the results for indexes look incorrect to me. The effective DDL of an index shouldn't contain the PK suffix inherited from the base table or parent view. (I can see it being useful to include such a thing in a SQL comment or something, but not the CREATE INDEX statement itself)

File: phoenix-tools/src/it/java/org/apache/phoenix/schema/SchemaExtractionToolIT.java [hunk 3]
```diff
@@ -160,12 +160,13 @@ public void testCreateViewIndexStatement() throws Exception {
         queries.add(createView);
         queries.add(createView1);
         queries.add(createIndexStatement);
+        String expected = "CREATE INDEX "+indexName+" ON " +childviewName +"(ID1, K, ID2, ID3 DESC) INCLUDE (V1)";
```
[the remark this answers is not in the document]
Comment (swaroopak) [reply phoenix-tools/src/it/java/org/apache/phoenix/schema/SchemaExtractionToolIT.java:163]: I tested this locally.
Effective DDL executed correctly and at sys cat level they both are exactly same including column ordering

**Comment by stoty** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m  4s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ 4.x Compile Tests _ |
| +0 :ok: |  mvndep  |   5m  7s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |   9m 19s |  4.x passed  |
| +1 :green_heart: |  compile  |   1m 32s |  4.x passed  |
| +1 :green_heart: |  checkstyle  |   1m  2s |  4.x passed  |
| +1 :green_heart: |  javadoc  |   1m  0s |  4.x passed  |
| +0 :ok: |  spotbugs  |   3m  0s |  phoenix-core in 4.x has 946 extant spotbugs warnings.  |
| +0 :ok: |  spotbugs  |   0m 45s |  phoenix-tools in 4.x has 3 extant spotbugs warnings.  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 17s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   6m  2s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m 32s |  the patch passed  |
| +1 :green_heart: |  javac  |   1m 32s |  the patch passed  |
| -1 :x: |  checkstyle  |   0m 45s |  phoenix-core: The patch generated 2 new + 719 unchanged - 3 fixed = 721 total (was 722)  |
| -1 :x: |  checkstyle  |   0m 15s |  phoenix-tools: The patch generated 34 new + 192 unchanged - 31 fixed = 226 total (was 223)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  javadoc  |   0m 57s |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   4m  5s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 211m 55s |  phoenix-core in the patch passed.  |
| +1 :green_heart: |  unit  |   4m 18s |  phoenix-tools in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 18s |  The patch does not generate ASF License warnings.  |
|  |   | 254m 27s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/2/artifact/yetus-general-check/output/Dockerfile |
| GITHUB PR | https://github.com/apache/phoenix/pull/1212 |
| JIRA Issue | PHOENIX-6271 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs hbaseanti checkstyle compile |
| uname | Linux 0820657ae601 4.15.0-112-generic #113-Ubuntu SMP Thu Jul 9 23:41:39 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev/phoenix-personality.sh |
| git revision | 4.x / c3f166e |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/2/artifact/yetus-general-check/output/diff-checkstyle-phoenix-core.txt |
| checkstyle | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/2/artifact/yetus-general-check/output/diff-checkstyle-phoenix-tools.txt |
|  Test Results | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/2/testReport/ |
| Max. process+thread count | 5094 (vs. ulimit of 30000) |
| modules | C: phoenix-core phoenix-tools U: . |
| Console output | https://ci-hadoop.apache.org/job/Phoenix/job/Phoenix-PreCommit-GitHub-PR/job/PR-1212/2/console |
| versions | git=2.7.4 maven=3.3.9 spotbugs=4.1.3 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java [hunk 4]
```diff
@@ -47,6 +47,7 @@
 
 import javax.annotation.Nullable;
 
+import com.google.common.base.Strings;
```
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/util/SchemaUtil.java:50]: Remember that you'll have to use the shaded version when you port to master.

File: phoenix-tools/src/it/java/org/apache/phoenix/schema/SchemaExtractionToolIT.java [hunk 5]
```diff
@@ -317,19 +321,104 @@ public void testCreateTableWithMultipleCFProperties() throws Exception {
...
+                + "ON MY_SCHEMA.MY_DATA_VIEW (TEXT1, DATE_TIME1 DESC, DOUBLE1)\n"
+                + "INCLUDE (CREATED_BY, CREATED_DATE)";
+        List<String> queries = new ArrayList<String>(){};
+
+        queries.add(table);
+        queries.add(view);
+        queries.add(index);
+
+        String expectedIndex = "CREATE INDEX MY_VIEW_INDEX "
+                + "ON MY_SCHEMA.MY_DATA_VIEW(TEXT1, DATE_TIME1 DESC, DOUBLE1, INT1)"
```
[the remark this answers is not in the document]
Comment (gjacoby126) [reply phoenix-tools/src/it/java/org/apache/phoenix/schema/SchemaExtractionToolIT.java:386]: Talked with @swaroopak about this offline, and she proved that this is OK, and added tests to that effect.

**Review by gjacoby126 — APPROVED** [review]
+1, thanks for the patch and adding the extra test assertions, @swaroopak

**Review by gokceni — APPROVED** [review]

**Comment by swaroopak** [comment]
Thank you for the review @gjacoby126 @gokceni

**Pull request merged**